### ISSUE TITLE
SRD Data

### ISSUE THREAD

**Issue by bgotink** [eepMoody/open5e#171]
I've been looking around on the web, and I couldn't find any good sources for the 5e SRD data. I'm not interested in a PDF or HTML, I need data files (e.g. JSON or YAML) containing the spells, feats etc. in a structured way.

I've tried the `make json`, but outputs the files in a way that can be used in a website. I would like something similar to (example in YAML because it's easier to write than JSON)

``` yaml
name: Acid Arrow
level: 2
school: evocation
ritual: false
casting time: 1 action
range: 90 ft
components:
  verbal: true
  somatic: true
  material: powdered rhubarb leaf and an adder's stomach
duration: instantaneous
description: '''
A shimmering green arrow streaks toward a target within range and bursts
in a spray of acid. Make a ranged spell attack against the target. On a
hit, the target takes 4d4 acid damage immediately and 2d4 acid damage at
the end of its next turn. On a miss, the arrow splashes the target with
acid for half as much of the initial damage and no damage at the end of
its next turn.
'''
at higher levels: '''
When you cast this spell using a spell slot of 3rd
level or higher, the damage (both initial and later) increases by 1d4
for each slot level above 2nd.
'''
```

This would allow for easy filtering on school, level, components, duration etc.

Is there any interest to implement something like this? In my opinion it would not only be a great addition to the open5e website if we could filter dynamically instead of only on first letter and on class, but it would allow for other projects to use this project as datasource (by publishing as npm module?).

**Comment by eepMoody** [comment]
Yeah, this is absolutely something we were hoping to do. Issue we ran into was figuring out how to parse the data we have into JSON or similar. (Once we get it into some sort of structured data, we figure it'll be easy to move into another type)

So far we were running up against issues with the Sphinx build process, but I've been debating moving our core build process out of Sphinx and basing our data elsewhere. (eg. constructing our Sphinx files from JSON snippets or similar)

Dream would be to have one source of data truth and build everything else from it, either from .rst > JSON or vice-versa.

If you have any thoughts, I'd be super interested to hear them!

**Comment by maxrothman** [comment]
To add to what @eepMoody said, this was one of our primary goals when starting Open5e ([our readme has said so since nearly the beginning of the project](https://github.com/eepMoody/open5e/commit/5422dc73f7abc8e590f3487f7978747b588383a2)), but but since this is a side project for all of us, getting momentum has been challenging. The last progress I made was in late June with a [data schema proposal](https://github.com/eepMoody/open5e/wiki/RFC:-Schemas) and some design of a [potential RST interface](https://github.com/eepMoody/open5e/wiki/WIP:-Directive-Interfaces) to structure the data, but I haven't been able to make significant progress since then.

On the other hand, you're the second person to ask about this since then, so I think we're quickly approaching critical mass to make this change happen. If you're interested in contributing, send me an email (on my profile) I'll do whatever I can to help you get up to speed. 

(Sidenote: how did we not already have an issue open for this‽)

**Comment by bgotink** [comment]
__TL;DR__: I'd be more than willing to help with this, though I'm unfamiliar with the data format and script language used in the project.
I expect it'll be difficult to find a schema that's structured enough to be useful but loose enough that it can describe all spells.

I've been toying around with an RST parser in NodeJS (I'm totally unfamiliar with the RST format and Python) and I got it to output a JSON with some information

```json
{
  "id": "word-of-recall",
  "name": "Word of Recall",
  "level": 6,
  "school": "conjuration",
  "ritual": false,
  "description": [
    "You and up to five willing creatures within 5 feet of you instantly teleport to a previously designated sanctuary. You and any creatures that teleport with you appear in the nearest unoccupied space to the spot you designated when you prepared your sanctuary (see below). If you cast this spell without first preparing a sanctuary, the spell has no effect.",
    "You must designate a sanctuary by casting this spell within a location, such as a temple, dedicated to or strongly linked to your deity. If you attempt to cast the spell in this manner in an area that isn't dedicated to your deity, the spell has no effect."
  ],
  "casting time": "1 action",
  "range": "5 feet",
  "components": {
    "verbal": true,
    "somatic": false,
    "material": false
  },
  "duration": "Instantaneous"
}
```

This example uses the format I described above. I just created it to alleviate my current spell list needs. The parser is finicky though, because I'm parsing (structured) text.

The big issue / discussion I see is not really the format (as long as it can easily be parsed into JSON/YAML/XML, it can be used as data source). I see more problems in the scheme, more specifically in the degree of structure the schema will impose on the data.
To show what I mean:

- Will the description just be text? That would mean we lose the ability to properly create references like

  ```rst
  :ref:`srd:restrained`
  ```

  or to properly format spells like Wish and Contagion which contain a list.  
  -> not structured enough limits possiblities
- Will we write the casting time in a structured format? From the wiki page:

  ```yaml
  casting_time:
      num: int
      unit: str
      ordinality: int #sorting val.  #name?
      (if unit: reaction)
      condition: str #e.g. when an ally within x feet is hit by an attack...
      (endif)
  ```

  If yes: how will we handle spells like Plant Growth, with two possible casting times?  
  -> too structured limits possiblities (or we'll need a lot of exceptions in the format)

**Comment by maxrothman** [comment]
From your repos, it seems like you've got experience with Ruby, C, and Java, and I have a lot of experience with Python, so I think between the two of us we'll be alright. If you're up for it, I am. 😃 

You're right on the money with the difficulties in creating a flexible enough schema. I wouldn't describe the schema in the wiki as more than a first pass, so your examples are really helpful. As I see it, there are two approaches we could take:

1. Keep playing with the JSON-style<sup>*</sup> schema and try to get it general enough to encompass the whole space of possible entities
  - Pro: more familiar for most probable API consumers
  - Pro: API has a well-defined schema, so parsers/clients will be more likely to be correct and bug-free
  - Con: harder/riskier to design, therefore increasing the likelihood of "design bugs", leading to more API versions, leading to more incorrect clients/parsers
    - We do have an existing corpus to test against, and it seems unlikely that WoTC will dump a ton of new entities that'll break our schema
2. Make the HTML semantic enough that it is itself the API
  - Con: less familiar for most probable API consumers
    - Though, the state of HTML scrapers/parsers is pretty advanced. JQuery or BeautifulSoup would make it pretty simple to just scrape the info you wanted out of the page if it was all tagged properly
  - Con: schema is by nature less structured, so parsers/clients are more likely to have bugs
    - We could distribute a test suite to help mitigate this
    - We could build tooling to create an XML schema by looking at the (X)HTML we already made, I think there might even already be tooling for that
  - Pro: easier to design (just tag stuff)

My feeling is that prioritizing API users by choosing option 1 would be the better choice if we can pull it off so I think it's worth trying to work with the JSON schema and seeing if we can make it work, but there's definitely a lot of design risk there.

To address your specific points:

1. Twitter's API handles the links-in-text problem pretty well. They do something like this:

        description: this is some text, and this should be a reference
        links:
        - rel: a reference
          href: /reference
          start: 20
          end: 50

2. Yeah, I'm not sure what to do about markup like lists and tables. I think we need to understand the usecase of consumers of the description field better.
3. I guess casting time could be a list? This is definitely a little squishy. I don't think we're aiming to make the API enable machines to understand/police a game, just to make it so they can put the right text in the right places and handle some simple rules stuff. A human is going to have to read the description to figure out what the spell does anyways.

Obviously there might be other approaches I haven't thought of, so if you have any ideas, I'd love to hear them.

---
<sup>*</sup>i.e. a schema that can be described by a combination of lists, maps, and scalars; so JSON, YAML, basically anything that will deserialize into native datastructures in most scripting languages.

**Comment by doxxon** [comment]
Why not use a database as your single source of truth and then move on defining an API?

**Comment by maxrothman** [comment]
@doxxon A couple points:

1. What would the schema of the database be? That just moves the problem around.
2. Open5e is currently a completely static site. It gets complied from RST to other formats (currently HTML, but hopefully an API format eventually). I don't see much benefit to adding a dynamic component that stores essentially static content.

**Comment by doxxon** [comment]
Making the SRD data accessible via API would open up a ton of opportunities
to developers.

While going to a database does push the problem around, it would push it
upstream.

OP's request was for a something that could make it easier to dynamic
filter, use as a data source, etc...

In any case, even a JSON format would be hugely helpful!!





On Tue, Oct 4, 2016 at 10:27 AM Max Rothman <notifications@github.com>
wrote:

@doxxon <https://github.com/doxxon> A couple points:

   1. What would the schema of the database be? That just moves the problem
   around.
   2. Open5e is currently a completely static site. It gets complied from
   RST to other formats (currently HTML, but hopefully an API format
   eventually). I don't see much benefit to adding a dynamic component that
   stores essentially static content.

—
You are receiving this because you were mentioned.


Reply to this email directly, view it on GitHub
<https://github.com/eepMoody/open5e/issues/171#issuecomment-251272855>, or mute
the thread

**Comment by maxrothman** [comment]
Can you describe your use case? I don't disagree that an API would be good
— that's the whole point of this issue — but I fail to see how adding a
database (I assume you mean a relational one) helps that goal, and I don't
understand your distinction between an API and a JSON format. As long as
the data is accessible in a machine-parsable format, isn't that an API?

Dynamic filtering would be a separate feature, which, though interesting, I
consider low priority given that the list of spells is small enough not to
require pagination. I'm happy to be convinced otherwise though.

**Comment by bgotink** [comment]
About the casting time example: I don't think anyone's going to build an AI, so a computer wouldn't need to understand what is going on. But there might be a usecase for being able to filter on casting time, suppose I'm writing a spellbook app and I want to allow users to quickly search their spellbook for spells they can cast in a reaction.

---

> Make the HTML semantic enough that it is itself the API
>   - Con: less familiar for most probable API consumers
>     - Though, the state of HTML scrapers/parsers is pretty advanced. JQuery or BeautifulSoup would make it pretty simple to just scrape the info you wanted out of the page if it was all tagged properly
>   - Con: schema is by nature less structured, so parsers/clients are more likely to have bugs
>     - We could distribute a test suite to help mitigate this
>     - We could build tooling to create an XML schema by looking at the (X)HTML we already made, I  think there might even already be tooling for that
>   - Pro: easier to design (just tag stuff)

I would not go for "we'll create a kinda structured page and then scrape it using jQuery" because then the HTML layout is fixed, and any update to the HTML will require updates in the scraper. There's absolutely no chance of properly layouting lists or tables in spell descriptions without an elaborate piece of code in the scraper to handle this.

There are several possibilities which make parsing easier though. We could look into semantic tags and attributes in the HTML, similar to w3's schema.org initiative using microdata or RFDa. ([interesting read](https://webdesign.tutsplus.com/articles/an-introduction-to-structured-data-markup--webdesign-8577)). I do not know whether any of that is possible in RST though.

The way I see it, if we store our data in JSON/YAML/whatever, we'll need to write a renderer to output HTML from these data files. If we write our data in the HTML, we'll need a parser to extract the data. Both the renderer and the parser are non-trivial problems, and both require a well thought out data schema.

---

Using a database doesn't solve the problem, it just moves the problem from "what kind of JSON schema will we use?" to "what kind of DB schema will we use?" These questions are equally hard.

I don't think that anyone is going to host a database of SRD data for other applications to consume. Neither do I believe that anyone is seriously considering consuming such a remote database, because you introduce an external source of failure in any application you build (do I as database consumer know that the person hosting the database will continue to pay for hosting? No)

**Comment by eepMoody** [comment]
+1 to avoiding a database. Given the static nature of the contents we host it feels like overkill.

I would push for building in JSON and rendering to html over the reverse, primarily because

**Comment by maxrothman** [comment]
@bgotink 

> But there might be a usecase for being able to filter on casting time, suppose I'm writing a spellbook app and I want to allow users to quickly search their spellbook for spells they can cast in a reaction.

Good point. What about something like this?

```
casting_time:
    options:    #name?
    - num: int
      unit: str
      (if unit: reaction)
      condition: str #e.g. when an ally within x feet is hit by an attack...
      (endif)
    - ...
    ordinality: int #sorting val.  #name?
```

By keeping the ordinality outside of the list, the spell remains globally sortable. Though, this begs the question of how spells with multiple casting times should be sorted.

> We could look into semantic tags and attributes in the HTML, similar to w3's schema.org initiative using microdata or RFDa. (interesting read). I do not know whether any of that is possible in RST though.

Also a good idea. Embedding markup in the text definitely gives us more display flexibility than embedding the text in markup. Sphinx/RST is actually great at this because it allows us to control the HTML output of any/all markup constructs. This approach has more or less equivalent tradeoffs as the text-in-markup HTML approach though.

---

@eepMoody 

> I would push for building in JSON and rendering to html/rst over the reverse...instead of providing people with a direct copy of our content and asking them write a parser

While I agree that it'll definitely be easier for API consumers to get started with a JSON format (see my pros/cons), that doesn't mean they won't need a client/parser. Every place the data spec has an `if` is a place where a client application would need to add logic. Also, is `$("#casting_time > #ordinality")` that much harder than `spell.casting_time.ordinality`?

> and force them to do more to just duplicate our site (by rendering it back into static HTML).

Given that the site is open-source and has a travisfile checked in that does automatic deployment to S3 and **still** nobody's duplicated the site (that we know of), I'm not particularly worried about that.

> eg. for the example of Plant Growth, they can do a search on "day" or "action" in the duration field, even if they're both just listed in a line.

I'm hesitant to push logic like that into clients, since doing so could constitute the need for a thicker "implicit client" (i.e. the code everyone needs to write to work with your API), hence more client bugs. As an API consumer, I would want the API to straightforwardly support common usecases (e.g. searching by casting time).

In general, I think we're not all on the same page as to what usecases we want to support and how we want to support them. It seems like you're pushing for minimizing time-to-value for API consumers, whereas I'm trying to minimize friction down the line incurred by altering the data format. Perhaps we should come up with some (\*gag*) user stories?

**Comment by eepMoody** [comment]
So my original intent in creating this was to solve the problem "I want to build an app for dnd, and I need a machine-readable source of the srd text".

I honestly think we've gone a bit overboard on trying to think of every way someone might want to work with the data. A simpler schema like what @bgotink posted is what I was hoping to find and then set out to build, and what I as a developer would pick if presented with both of these options. It allows me to choose how I might subdivide or parse those pieces of info, and puts them in a useful format out of the gate.

I do think we should push work on the developers, by which I really mean that we should _allow_ them to parse and manipulate the text in the way they wish instead of presenting an opinionated structure.

If we then get feedback that people want more structure, then I think we build another endpoint that provides a different structure. Exporting it two different ways won't be much work, picking the schema seems to be the harder part.

**Comment by eepMoody** [comment]
@bgotink could you see about getting that parsed data from your first comment working in our build and make a branch so we can see what it might look like applied to the whole list? I'll get you access so you can do it right in the repo here.

**Comment by bgotink** [comment]
I've pushed the `poc-json-schema` branch. As stated above, it's written in js, so you'll need Node 4 or later.

```bash
npm i
npm run extract-json # or make datajson
```

Note: A bunch of spells are missing due to seikichi/restructured#2

**Comment by maxrothman** [comment]
> Exporting it two different ways won't be much work, picking the schema seems to be the harder part.

Fair point. Multiple APIs are pretty painless for us.

It sounds like there's a third proposal here: a minimal, semi-structured JSON format where most values are strings. I think there's merit to that approach, and it wouldn't be significantly harder to have Sphinx output that compared to any of the other formats. Let's definitely consider that alongside the others, and gathering feedback on it as @eepMoody suggested will not only help us weigh it against the other options, but also give us some concrete user stories.

**Comment by maxrothman** [comment]
@bgotink does the result of `make datajson` have a consistent schema, and if so, can you put up a wiki page so we have a central place to work on it?

Also, looking at that bug, I wonder if it's because `:ref:` is a Sphinx-specific structure role. It's not built into RST.

**Comment by bgotink** [comment]
> @bgotink does the result of make datajson have a consistent schema, and if so, can you and @eepMoody work on a wiki page so we have a central place to work on it?

Yeah, the schema is consistent. I've added the current state of the format to [the Schema page](https://github.com/eepMoody/open5e/wiki/RFC:-Schemas#simplified-spell).

> Also, looking at that bug, I wonder if it's because :ref: is a Sphinx-specific structure role. It's not built into RST.

It looks like a regression, the previous version of the npm module could handle the references perfectly.

**Comment by maxrothman** [comment]
We've made https://github.com/eepMoody/open5e/wiki/In-Progress:-Schemas to collect some of the key parts of this discussion into a single place. Feel free to add to it!

Also for important and I-promise-completely-reasonable reasons, @bgotink you should send me an email. (I really wish gh had private messages.)

**Comment by eepMoody** [comment]
I did some playing with the poc and some potential additions to the schema in the form of useful metadata. Examples I sketched out: Damage types dealt by the spell, damage rolls (from which approximate damage could be calculated), and concentration as a boolean value.

Other interesting metadata possibilities along a similar route that would be easy to implement: Conditions mentioned, healing, etc.

This is starting to look like an interesting possibility to create a middle ground between the two options presented: instead of attempting to do a 1:1 conversion of rst to JSON elements, we could provide the strings in simple format plus useful metadata. A big win there is that adding new metadata wouldn't be a breaking change, so we could progressively add extra info as people require it.

**Comment by maxrothman** [comment]
I think this is a great expansion on the "Lite JSON" idea. I have some concerns about client usability, mostly around how API users would know which fields to look for. If we have too many, it'll mean a giant pile of if-statements for them. Maybe we could have 2 sub-objects, one with "required" fields, and one with "optional" fields? Though that's starting to sound like a new proposal. I think it would also be useful to use objects instead of strings for all of the metadata so they can show where they appear in the text. (e.g. `damageTypes: { [{name: "piercing", start: 38, end: 45, field: "description"}, ...] }` It's a little clunky and would be more natural in an XML format, but I think it'd be useful.

One thing we haven't talked at all about is how we'd handle rich markup in any of the JSON formats, e.g. tables, lists, etc. Any ideas?

For the record, I really like how we started with 2 proposals and now we maybe have 4. This has been an awesome discussion.

**Comment by eepMoody** [comment]
Maybe some sort of "meta" group would be the best way to include all the optional fields? Make sure we put all the absolutely necessary data (the SRD text) in the main section, and then all our extras which are more optional in `meta`?

My thinking is that we want to support the most basic use-case (displaying SRD info and being able to search against it) without any logic. Anyone taking advantage of our additional metadata to do other cool stuff is going to be in a position where creating logic to deal with it is expected.

**Comment by maxrothman** [comment]
👍 to that general idea, though I'd call the field `fields` or `extra` or something. `meta` is usually used for metadata about the request itself, e.g. pagination, HATEOAS links, etc.

**Comment by eepMoody** [comment]
This sort of content actually seems in keeping with the sort of thing specified for `meta` tags in the JSON API spec. http://jsonapi.org/format/#document-meta

"...a meta member can be used to include non-standard meta-information."

Some of the examples they use are authors, copyright dates, etc. Links and pagination have their own structures.

**Comment by maxrothman** [comment]
Sorry yes, my examples were not great, but I would still argue that the content we're discussing is actual content, not metadata *about* the content.

**Comment by eepMoody** [comment]
I'd disagree for most of it. Damage types references, classes that can cast or access something, conditions involved, etc. are basically content tags, which belong in meta. The rolls may be the outlier that belongs in some other place, but `meta` is where I'd look for all the others.

I'm also thinking beyond spells specifically. What about monsters or class abilities that deal various types of damage or have certain movement speeds, can inflict certain conditions, etc. This would be a nice way to put the same type of info in the same place across various things (and put it in a standard location that's used in most APIs).

**Comment by eepMoody** [comment]
Looks like discussion died down a bit, and I think there's a good first pass ready to go. 

V1 will be a JSON API that looks like this unless someone spots a major issue with it, and we'll see what the feedback is once we roll it out. 

```YAML
id: string # lower case hyphenated (aka kebab-case)
name: string # human readable name
level: integer # values 0 - 9, 0 signifies cantrip
school: string # lowercase, enum with possible values "conjuration", "evocation", "necromancy" etc.
ritual: boolean
casting time: string # possibly string[] for different casting times?
concentration: boolean
duration: string # possibly string[] for different durations?
range: string # possibly string[] for different ranges?
components:
  verbal: boolean
  somatic: boolean
  material: false | string
description: string[] # description split in paragraphs
at higher levels: string[] | null # paragraphs or null, possibly use empty array instead of
meta:
  damageTypes: string[] | null #with possible values "fire", "psychic", "necrotic", etc.
  classes: string[] | null #with possible values "druid", "wizard", "bard", etc.
  rolls: object[]
    index
      text: string #the roll text
      from: int #start position of string
      to: int #end position of string
      field: string #from properties in main section
  ...
links:
  self:
    href: string #link to location on our site
```

If someone can figure out how to get reference links or other info from our site in in before we're ready I to go let's do it, otherwise I think this will probably be enough that people are excited to build stuff with it (I know I already am).

@bgotink what do you think is the best way to start turning the PoC you built into something a bit cleaner?

**Comment by bgotink** [comment]
Unless you plan on moving everything to NodeJS, I'd start by rewriting the parser in python. If there's an intermediary data format that won't lead to a lot of "if this element is a paragraph, do foo; else ...", I'd use that instead of the RST files.

**Comment by maxrothman** [comment]
FWIW the docutils node tree is serialized on-disk as native python objects using the pickle module, so you could just deserialize that and look at it.

**Comment by eepMoody** [comment]
@bgotink Personally I don't actually know how to work in Python at all, so the NodeJS version is much more accessible to me. I think the RST is a bit too unstructured as-is to avoid doing a bit of text parsing no matter what we work in. We were investigating creating some more structured domains in Sphinx to solve that problem, but it's turned out to be much more involved that it looked like it would be.

I dunno about rebuilding _everything_ in NodeJS, but if there's an equally good solution that doesn't require a huge amount of work I think it'd be worth considering. 

We haven't found anything that does as good a job at creating referenced documentation as well as Sphinx does, but maybe there are solutions out there that we missed the first time around that would hit all the useful functions and be easier to expand/modify? Sphinx gave us a LOT of great stuff out of the box, but expanding functionality on it has been very tough. Search and data structure in particular have been difficult to improve on, and other libraries I've found seem to be really mixed in quality or have been tough to find up-to-date versions of.

**Comment by bgotink** [comment]
Parsing the doctree file for Aid loads the following XML document:

```xml
<document source="/path/to/open5e/source/Spellcasting/spells_a-z/a/aid.rst">
  <target refid="srd-aid"/>
  <section ids="aid srd-aid" names="aid srd:aid">
    <title>
      Aid
    </title>
    <section ids="nd-level-abjuration" names="2nd-level\ abjuration">
      <title>
        2nd-level abjuration
      </title>
      <paragraph>
        <strong>
          Casting Time:
        </strong>
         1 action
      </paragraph>
      <paragraph>
        <strong>
          Range:
        </strong>
         30 feet
      </paragraph>
      <paragraph>
        <strong>
          Components:
        </strong>
         V, S, M (a tiny strip of white cloth)
      </paragraph>
      <paragraph>
        <strong>
          Duration:
        </strong>
         8 hours
      </paragraph>
      <paragraph>
        Your spell bolsters your allies with toughness and resolve. Choose up to
three creatures within range. Each target's hit point maximum and
current hit points increase by 5 for the duration.
      </paragraph>
      <paragraph>
        <strong>
          At Higher Levels.
        </strong>
         When you cast this spell using a spell slot of 3rd
level or higher, a target's hit points increase by an additional 5 for
each slot level above 2nd.
      </paragraph>
    </section>
  </section>
</document>
```

That's not really more useful than parsing the RST files directly.

(source:

```python
from pickle import Unpickler;
from yattag import indent;

f = open('a/aid.doctree', 'r')
result = Unpickler(f).load();
print indent(str(result), indent_text = True);
```
)

I'll see whether I can reproduce my previous JSON with this data (should not be a problem) in Python (I can always ask my python-loving partner for help).

**Comment by maxrothman** [comment]
Not sure if it makes a difference, but when you call `str()` on the unpickled file, what you're getting is the string representation of the object. The object itself is actually a tree of RST node objects. You can poke around and see what properties/methods they have by calling `dir()` to list them and `help()` to see their documentation. I know there's a `traverse` method that might be helpful. 

Alternatively you could injest the entire thing as xml using the etree module which has pretty powerful selectors and filtering.

**Comment by eepMoody** [comment]
@bgotink yeah that's about what I expected. I figure there's no way any transformation is actually going to add more structure than exists in the original, at best it will be lossless and provide a slightly easier syntax.

I'm a little fuzzy on how I'd actually call a parsing function on data in this format, but if you want to sketch one out and check it in I should be able to help with building out all the actual logic?

**Comment by HopScotch47** [comment]
Are you guys still working on/thinking about this?
I am yet another person looking for the srd stuff in json to consume in my little dnd encounter manager

Happy to help if I can, but there is absolutely no Python skills here, I am a .net and js kinda guy

**Comment by brettstack** [comment]
+1 came here also looking for JSON data. Any update? :)

**Comment by maxrothman** [comment]
Thanks for the interest! We did make a little bit of progress at one point in taking some existing JSON data from another source and adapting it for use on this site, but since [D&D Beyond](https://www.dndbeyond.com/) was released and it does everything we tried to do, but better and fully supported by WotC, we've mostly stopped development on this project. They've also stated that APIs are on their roadmap, so I'm watching closely.

If y'all are still interested in working on JSON data sooner than when WotC gets to it, we're happy to give you push access to the repo.
